### PR TITLE
69th PR Nice - Fix Python 3.10 asyncio.gather exception

### PR DESCRIPTION
**Pull request by Pirulax**
In Py 3.10 `asyncio.gather` wont' take the loop in as a kwarg, but rather use `_events._get_event_loop()` which, since there's no loop set, will raise an exception.
But this only happens if there are no tasks to be cancelled. If there are everything is going to be fine.
For simplicity I put the whole thing under the `if` statement, because running `asyncio.gather` with no tasks would have no effect on older versions either.

I suggest adding a regression test for this, at some point.
Just please merge it ASAP, because I don't like having modified vendor code :D

Also, thanks for the awesome lib, I'm thankful I got rid of the boilerplate.

### CODE REVIEW

**Comment by Pirulax** [comment]
Perhaps, one other solution that might work, would be to call `asyncio.set_event_loop`.
That way user code (eg.: `shutdown_callback`) could use `asyncio.gather` as well (Which it can't currently).

**Comment by cjrh** [comment]
Thanks for this. I have some leave coming up so I'll look at this in the next few days 👍🏼

**Comment by Pirulax** [comment]
Thank you!

File: aiorun.py [hunk 1]
```diff
@@ -318,22 +318,27 @@ def sep():
...
+        # the gathered group will actually be complete. You need to
+        # enable this with the ``return_exceptions`` flag.
+        if sys.version_info < (3, 10):
+            group = gather(*tasks, *do_not_cancel, return_exceptions=True, loop=loop)
+        else:
+            group = gather(*tasks, *do_not_cancel, return_exceptions=True)
+        logger.info("Running pending tasks till complete")
+        # TODO: obtain all the results, and log any results that are exceptions
+        # other than CancelledError. Will be useful for troubleshooting.
+        loop.run_until_complete(group)
```
Comment (cjrh) [inline aiorun.py:341]: Thanks again for raising the issue.  I looked at it and I think the following is a better change overall (and I should have done it this way the first time round):

```suggestion
    async def wait_for_cancelled_tasks():
        return await gather(*tasks, *do_not_cancel, return_exceptions=True)

    if tasks or do_not_cancel:
        logger.info("Running pending tasks till complete")
        # TODO: obtain all the results, and log any results that are exceptions
        # other than CancelledError. Will be useful for troubleshooting.
        loop.run_until_complete(wait_for_cancelled_tasks())
```

What are your thoughts on this?

**Review by cjrh — CHANGES_REQUESTED** [review]

File: aiorun.py (hunk 1 above)
Comment (Pirulax) [reply aiorun.py:341]: Sure, seems good to me.
I guess now you don't need to pass `loop` to `gather` even in Py < 3.10 ?

Comment (cjrh) [reply aiorun.py:341]: Yes exactly.  Are you happy for me to make a PR myself and close this one, or do you intend to change yours (this PR) to look like above?

Comment (Pirulax) [reply aiorun.py:341]: Should I just commit your suggestion?

**Review by cjrh — APPROVED** [review]

**Comment by cjrh** [comment]
Thanks!

**Comment by Pirulax** [comment]
Thank you too!

**Comment by Pirulax** [comment]
btw, this killed CI for 3.5, and 3.6..

**Pull request merged**